python/ray/_raylet.pyx

```diff
@@ -708,289 +713,302 @@ cdef execute_task(
-                            dynamic_returns[0][idx].first.Binary(),
-                            caller_address.SerializeAsString(),
-                        ))
-                    # Swap out the generator for an ObjectRef generator.
-                    outputs = (ObjectRefGenerator(dynamic_refs), )
-
-                # TODO(swang): For generator tasks, iterating over outputs will
-                # actually run the task. We should run the usual handlers for
-                # task cancellation, retrying on application exception, etc. for
-                # all generator tasks, both static and dynamic.
-                core_worker.store_task_outputs(
-                    worker, outputs,
-                    returns)
-        except Exception as error:
+                        )
+                        if (
+                            is_retryable_error[0]
+                            and core_worker.get_current_task_retry_exceptions()
+                        ):
+                            logger.info("Task failed with retryable exception:"
+                                        " {}.".format(
+                                            core_worker.get_current_task_id()),
+                                        exc_info=True)
+                        else:
+                            logger.debug("Task failed with unretryable exception:"
+                                         " {}.".format(
+                                            core_worker.get_current_task_id()),
+                                         exc_info=True)
+                        raise e
+                    if returns[0].size() == 1 and not inspect.isgenerator(outputs):
+                        # If there is only one return specified, we should return
+                        # all return values as a single object.
+                        outputs = (outputs,)
+                if (<int>task_type == <int>TASK_TYPE_ACTOR_CREATION_TASK):
+                    # Record actor repr via :actor_name: magic token in the log.
+                    #
+                    # (Phase 2): after `__init__` finishes, we override the
+                    # log prefix with the full repr of the actor. The log monitor
+                    # will pick up the updated token.
+                    if (hasattr(actor_class, "__ray_actor_class__") and
+                            (actor_class.__ray_actor_class__.__repr__
+                             != object.__repr__)):
+                        actor_magic_token = "{}{}\n".format(
+                            ray_constants.LOG_PREFIX_ACTOR_NAME, repr(actor))
+                        # Flush on both stdout and stderr.
+                        print(actor_magic_token, end="")
+                        print(actor_magic_token, file=sys.stderr, end="")
+
+                # Check for a cancellation that was called when the function
+                # was exiting and was raised after the except block.
+                task_exception = True
```

Comment:
Hmm I was thinking no since I didn't think we'd be executing any other Python statements that would result in the interpreter checking signals, but that's a brittle assumption. Yeah we should probably have another `PyErr_CheckSignals()` after outputs are stored!